components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -93,6 +94,49 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes and reads here.
+    protected readonly contentCache = new Map<string, string>();
+    protected updatedSinceLastWrite = new Set<string>();
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.add(resource.uri);
+    }
+
+    protected async cachedRead(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+
+        const content = await this.getServer().getUserStorageResource(req);
+        this.contentCache.set(req.uri, content);
+        return content;
+    }
+
+    protected startPeriodicCacheWriter(interval: number) {
+        const performWrites = async () => {
+            const writeStart = Date.now();
+
+            const writtenUris = this.updatedSinceLastWrite;
+            this.updatedSinceLastWrite = new Set<string>();
+
+            try {
```

Comment:
Should not try-catch wrap individual calls to `updateUserStorageResource` to avoid failing for one URI to break all others?

Reply:
The idea was if one fails, the other will fail as well. But will pick up your suggestion from below: Re-submitting URIs makes sense.

Reply:
I heard it happens sometime that content is too big to store in the database, not sure whether it fails this request then. cc @AlexTugarev ?